staging/src/k8s.io/client-go/rest/request.go

```diff
@@ -1339,17 +1357,21 @@ func truncateBody(body string) string {
 	return body[:max] + fmt.Sprintf(" [truncated %d chars]", len(body)-max)
 }
 
-// glogBody logs a body output that could be either JSON or protobuf. It explicitly guards against
+// logBody logs a body output that could be either JSON or protobuf. It explicitly guards against
 // allocating a new string for the body output unless necessary. Uses a simple heuristic to determine
 // whether the body is printable.
-func glogBody(prefix string, body []byte) {
-	if klogV := klog.V(8); klogV.Enabled() {
+//
+// It needs to be called by all functions which send or receive the data.
+func logBody(ctx context.Context, callDepth int, prefix string, body []byte) {
+	logger := klog.FromContext(ctx)
+	if loggerV := logger.V(8); loggerV.Enabled() {
+		loggerV := loggerV.WithCallDepth(callDepth)
 		if bytes.IndexFunc(body, func(r rune) bool {
 			return r < 0x0a
 		}) != -1 {
-			klogV.Infof("%s:\n%s", prefix, truncateBody(hex.Dump(body)))
+			loggerV.Info(prefix, "body", truncateBody(logger, hex.Dump(body)))
```

Comment:
how does `kubectl get --v=9` regress with this?

Reply:
Before:
```console
$ _output/bin/kubectl get -v=9 nodes
I0903 10:45:29.186147  411090 loader.go:395] Config loaded from file:  /var/run/kubernetes/admin.kubeconfig
I0903 10:45:29.193442  411090 round_trippers.go:466] curl -v -XGET  -H "Accept: application/json;as=Table;v=v1;g=meta.k8s.io,application/json;as=Table;v=v1beta1;g=meta.k8s.io,application/json" -H "User-Agent: kubectl/v1.32.0 (linux/amd64) kubernetes/f0307ee" 'https://localhost:6444/api/v1/nodes?limit=500'
I0903 10:45:29.194091  411090 round_trippers.go:495] HTTP Trace: DNS Lookup for localhost resolved to [{::1 } {127.0.0.1 }]
I0903 10:45:29.194369  411090 round_trippers.go:510] HTTP Trace: Dial to tcp:[::1]:6444 succeed
I0903 10:45:29.208429  411090 round_trippers.go:553] GET https://localhost:6444/api/v1/nodes?limit=500 200 OK in 14 milliseconds
I0903 10:45:29.208460  411090 round_trippers.go:570] HTTP Statistics: DNSLookup 0 ms Dial 0 ms TLSHandshake 10 ms ServerProcessing 3 ms Duration 14 ms
I0903 10:45:29.208479  411090 round_trippers.go:577] Response Headers:
I0903 10:45:29.208496  411090 round_trippers.go:580]     Cache-Control: no-cache, private
I0903 10:45:29.208510  411090 round_trippers.go:580]     Content-Type: application/json
I0903 10:45:29.208520  411090 round_trippers.go:580]     X-Kubernetes-Pf-Flowschema-Uid: f6e769a1-a9e5-4aac-b577-5caac5bb2904
I0903 10:45:29.208530  411090 round_trippers.go:580]     X-Kubernetes-Pf-Prioritylevel-Uid: 37118ad8-6c89-4219-94d6-451d8a78dae1
I0903 10:45:29.208540  411090 round_trippers.go:580]     Date: Tue, 03 Sep 2024 08:45:29 GMT
I0903 10:45:29.208549  411090 round_trippers.go:580]     Audit-Id: 7d9be584-08b5-447b-ac71-f608d597eb59
I0903 10:45:29.208805  411090 request.go:1352] Response Body: {"kind":"Table","apiVersion":"meta.k8s.io/v1","metadata":{"resourceVersion":"2657"},"columnDefinitions":[{"name":"Name","type":"string","format":"name","description":"Name must be unique within a namespace. Is required when creating resources, although some resources may allow a client to request the generation of an appropriate name automatically. Name is primarily intended for creation idempotence and configuration definition. Cannot be updated. More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names#names","priority":0},{"name":"Status","type":"string","format":"","description":"The status of the node","priority":0},{"name":"Roles","type":"string","format":"","description":"The roles of the node","priority":0},{"name":"Age","type":"string","format":"","description":"CreationTimestamp is a timestamp representing the server time when this object was created. It is not guaranteed to be set in happens-before order across separate operations. Clients may not set this value. It is represented in RFC3339 form and is in UTC.\n\nPopulated by the system. Read-only. Null for lists. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata","priority":0},{"name":"Version","type":"string","format":"","description":"Kubelet Version reported by the node.","priority":0},{"name":"Internal-IP","type":"string","format":"","description":"List of addresses reachable to the node. Queried from cloud provider, if available. More info: https://kubernetes.io/docs/concepts/nodes/node/#addresses Note: This field is declared as mergeable, but the merge key is not sufficiently unique, which can cause data corruption when it is merged. Callers should instead use a full-replacement patch. See https://pr.k8s.io/79391 for an example. Consumers should assume that addresses can change during the lifetime of a Node. However, there are some exceptions where this may not be possible, such as Pods that inherit a Node's address in its own status or consumers of the downward API (status.hostIP).","priority":1},{"name":"External-IP","type":"string","format":"","description":"List of addresses reachable to the node. Queried from cloud provider, if available. More info: https://kubernetes.io/docs/concepts/nodes/node/#addresses Note: This field is declared as mergeable, but the merge key is not sufficiently unique, which can cause data corruption when it is merged. Callers should instead use a full-replacement patch. See https://pr.k8s.io/79391 for an example. Consumers should assume that addresses can change during the lifetime of a Node. However, there are some exceptions where this may not be possible, such as Pods that inherit a Node's address in its own status or consumers of the downward API (status.hostIP).","priority":1},{"name":"OS-Image","type":"string","format":"","description":"OS Image reported by the node from /etc/os-release (e.g. Debian GNU/Linux 7 (wheezy)).","priority":1},{"name":"Kernel-Version","type":"string","format":"","description":"Kernel Version reported by the node from 'uname -r' (e.g. 3.16.0-0.bpo.4-amd64).","priority":1},{"name":"Container-Runtime","type":"string","format":"","description":"ContainerRuntime Version reported by the node through runtime remote API (e.g. containerd://1.4.2).","priority":1}],"rows":[{"cells":["127.0.0.1","Ready","\u003cnone\u003e","120m","v1.32.0-alpha.0.400+f0307eefb87bde-dirty","127.0.0.1","\u003cnone\u003e","Debian GNU/Linux 12 (bookworm)","6.1.0-18-amd64","cri-o://1.30.0"],"object":{"kind":"PartialObjectMetadata","apiVersion":"meta.k8s.io/v1","metadata":{"name":"127.0.0.1","uid":"7523429e-2e31-467f-98e2-1ab69b48c103","resourceVersion":"2634","creationTimestamp":"2024-09-03T06:45:10Z","labels":{"beta.kubernetes.io/arch":"amd64","beta.kubernetes.io/os":"linux","kubernetes.io/arch":"amd64","kubernetes.io/hostname":"127.0.0.1","kubernetes.io/os":"linux"},"annotations":{"node.alpha.kubernetes.io/ttl":"0","volumes.kubernetes.io/controller-managed-attach-detach":"true"},"managedFields":[{"manager":"test-client-A","operation":"Apply","apiVersion":"v1","time":"2024-09-03T06:48:38Z","fieldsType":"FieldsV1","fieldsV1":{"f:spec":{"f:taints":{"k:{\"effect\":\"NoSchedule\",\"key\":\"tainted-by-client-a\"}":{"f:value":{}}}}}},{"manager":"test-client-B","operation":"Apply","apiVersion":"v1","time":"2024-09-03T06:48:38Z","fieldsType":"FieldsV1","fieldsV1":{"f:spec":{"f:taints":{"k:{\"effect\":\"NoSchedule\",\"key\":\"tainted-by-client-b\"}":{"f:value":{}}}}}},{"manager":"kube-controller-manager","operation":"Update","apiVersion":"v1","time":"2024-09-03T06:45:10Z","fieldsType":"FieldsV1","fieldsV1":{"f:metadata":{"f:annotations":{"f:node.alpha.kubernetes.io/ttl":{}}}}},{"manager":"kubelet","operation":"Update","apiVersion":"v1","time":"2024-09-03T06:45:10Z","fieldsType":"FieldsV1","fieldsV1":{"f:metadata":{"f:annotations":{".":{},"f:volumes.kubernetes.io/controller-managed-attach-detach":{}},"f:labels":{".":{},"f:beta.kubernetes.io/arch":{},"f:beta.kubernetes.io/os":{},"f:kubernetes.io/arch":{},"f:kubernetes.io/hostname":{},"f:kubernetes.io/os":{}}}}},{"manager":"kubelet","operation":"Update","apiVersion":"v1","time":"2024-09-03T08:44:17Z","fieldsType":"FieldsV1","fieldsV1":{"f:status":{"f:conditions":{"k:{\"type\":\"DiskPressure\"}":{"f:lastHeartbeatTime":{}},"k:{\"type\":\"MemoryPressure\"}":{"f:lastHeartbeatTime":{}},"k:{\"type\":\"PIDPressure\"}":{"f:lastHeartbeatTime":{}},"k:{\"type\":\"Ready\"}":{"f:lastHeartbeatTime":{},"f:lastTransitionTime":{},"f:message":{},"f:reason":{},"f:status":{}}},"f:images":{},"f:runtimeHandlers":{}}},"subresource":"status"}]}}}]}

$ $ _output/bin/kubectl create -v=9 -f test/e2e/dra/test-driver/deploy/example/resourceclaim.yaml 
...
I0903 10:47:50.375117  413646 request.go:1352] Request Body: {"apiVersion":"resource.k8s.io/v1alpha3","kind":"ResourceClaim","metadata":{"name":"example","namespace":"default"},"spec":{"devices":{"requests":[{"deviceClassName":"example","name":"req-0"}]}}}
I0903 10:47:50.375225  413646 round_trippers.go:466] curl -v -XPOST  -H "Accept: application/json" -H "Content-Type: application/json" -H "User-Agent: kubectl/v1.32.0 (linux/amd64) kubernetes/f0307ee" 'https://localhost:6444/apis/resource.k8s.io/v1alpha3/namespaces/default/resourceclaims?fieldManager=kubectl-create&fieldValidation=Strict'
I0903 10:47:50.383587  413646 round_trippers.go:553] POST https://localhost:6444/apis/resource.k8s.io/v1alpha3/namespaces/default/resourceclaims?fieldManager=kubectl-create&fieldValidation=Strict 201 Created in 8 milliseconds
I0903 10:47:50.383636  413646 round_trippers.go:570] HTTP Statistics: GetConnection 0 ms ServerProcessing 8 ms Duration 8 ms
I0903 10:47:50.383653  413646 round_trippers.go:577] Response Headers:
I0903 10:47:50.383673  413646 round_trippers.go:580]     Cache-Control: no-cache, private
I0903 10:47:50.383684  413646 round_trippers.go:580]     Content-Type: application/json
I0903 10:47:50.383697  413646 round_trippers.go:580]     X-Kubernetes-Pf-Flowschema-Uid: f6e769a1-a9e5-4aac-b577-5caac5bb2904
I0903 10:47:50.383708  413646 round_trippers.go:580]     X-Kubernetes-Pf-Prioritylevel-Uid: 37118ad8-6c89-4219-94d6-451d8a78dae1
I0903 10:47:50.383720  413646 round_trippers.go:580]     Content-Length: 577
I0903 10:47:50.383729  413646 round_trippers.go:580]     Date: Tue, 03 Sep 2024 08:47:50 GMT
I0903 10:47:50.383743  413646 round_trippers.go:580]     Audit-Id: 2d4598a8-f4eb-4c10-a141-7fd7b4037242
I0903 10:47:50.383812  413646 request.go:1352] Response Body: {"kind":"ResourceClaim","apiVersion":"resource.k8s.io/v1alpha3","metadata":{"name":"example","namespace":"default","uid":"2a2ca5e9-d612-4c1d-b461-5ea8a4f77058","resourceVersion":"2704","creationTimestamp":"2024-09-03T08:47:50Z","managedFields":[{"manager":"kubectl-create","operation":"Update","apiVersion":"resource.k8s.io/v1alpha3","time":"2024-09-03T08:47:50Z","fieldsType":"FieldsV1","fieldsV1":{"f:spec":{"f:devices":{"f:requests":{}}}}}]},"spec":{"devices":{"requests":[{"name":"req-0","deviceClassName":"example","allocationMode":"ExactCount","count":1}]}},"status":{}}
resourceclaim.resource.k8s.io/example created

```

After:
```console
$ _output/bin/kubectl get -v=9 nodes
I0903 10:51:03.911741  419139 loader.go:395] Config loaded from file:  /var/run/kubernetes/admin.kubeconfig
I0903 10:51:03.920344  419139 helper.go:113] "Request Body" body=""
I0903 10:51:03.920464  419139 round_trippers.go:466] curl -v -XGET  -H "Accept: application/json;as=Table;v=v1;g=meta.k8s.io,application/json;as=Table;v=v1beta1;g=meta.k8s.io,application/json" -H "User-Agent: kubectl/v1.32.0 (linux/amd64) kubernetes/7dbe980" 'https://localhost:6444/api/v1/nodes?limit=500'
I0903 10:51:03.921130  419139 round_trippers.go:495] HTTP Trace: DNS Lookup for localhost resolved to [{::1 } {127.0.0.1 }]
I0903 10:51:03.921405  419139 round_trippers.go:510] HTTP Trace: Dial to tcp:[::1]:6444 succeed
I0903 10:51:03.932978  419139 round_trippers.go:553] GET https://localhost:6444/api/v1/nodes?limit=500 200 OK in 12 milliseconds
I0903 10:51:03.933023  419139 round_trippers.go:570] HTTP Statistics: DNSLookup 0 ms Dial 0 ms TLSHandshake 8 ms ServerProcessing 3 ms Duration 12 ms
I0903 10:51:03.933040  419139 round_trippers.go:577] Response Headers:
I0903 10:51:03.933056  419139 round_trippers.go:580]     Date: Tue, 03 Sep 2024 08:51:03 GMT
I0903 10:51:03.933066  419139 round_trippers.go:580]     Audit-Id: 756dcd6f-ba68-485c-9fa6-b9768915671b
I0903 10:51:03.933076  419139 round_trippers.go:580]     Cache-Control: no-cache, private
I0903 10:51:03.933087  419139 round_trippers.go:580]     Content-Type: application/json
I0903 10:51:03.933099  419139 round_trippers.go:580]     X-Kubernetes-Pf-Flowschema-Uid: f6e769a1-a9e5-4aac-b577-5caac5bb2904
I0903 10:51:03.933108  419139 round_trippers.go:580]     X-Kubernetes-Pf-Prioritylevel-Uid: 37118ad8-6c89-4219-94d6-451d8a78dae1
I0903 10:51:03.933264  419139 helper.go:113] "Response Body" body=<
	{"kind":"Table","apiVersion":"meta.k8s.io/v1","metadata":{"resourceVersion":"2770"},"columnDefinitions":[{"name":"Name","type":"string","format":"name","description":"Name must be unique within a namespace. Is required when creating resources, although some resources may allow a client to request the generation of an appropriate name automatically. Name is primarily intended for creation idempotence and configuration definition. Cannot be updated. More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names#names","priority":0},{"name":"Status","type":"string","format":"","description":"The status of the node","priority":0},{"name":"Roles","type":"string","format":"","description":"The roles of the node","priority":0},{"name":"Age","type":"string","format":"","description":"CreationTimestamp is a timestamp representing the server time when this object was created. It is not guaranteed to be set in happens-before order across separate operations. Clients may not set this value. It is represented in RFC3339 form and is in UTC.\n\nPopulated by the system. Read-only. Null for lists. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata","priority":0},{"name":"Version","type":"string","format":"","description":"Kubelet Version reported by the node.","priority":0},{"name":"Internal-IP","type":"string","format":"","description":"List of addresses reachable to the node. Queried from cloud provider, if available. More info: https://kubernetes.io/docs/concepts/nodes/node/#addresses Note: This field is declared as mergeable, but the merge key is not sufficiently unique, which can cause data corruption when it is merged. Callers should instead use a full-replacement patch. See https://pr.k8s.io/79391 for an example. Consumers should assume that addresses can change during the lifetime of a Node. However, there are some exceptions where this may not be possible, such as Pods that inherit a Node's address in its own status or consumers of the downward API (status.hostIP).","priority":1},{"name":"External-IP","type":"string","format":"","description":"List of addresses reachable to the node. Queried from cloud provider, if available. More info: https://kubernetes.io/docs/concepts/nodes/node/#addresses Note: This field is declared as mergeable, but the merge key is not sufficiently unique, which can cause data corruption when it is merged. Callers should instead use a full-replacement patch. See https://pr.k8s.io/79391 for an example. Consumers should assume that addresses can change during the lifetime of a Node. However, there are some exceptions where this may not be possible, such as Pods that inherit a Node's address in its own status or consumers of the downward API (status.hostIP).","priority":1},{"name":"OS-Image","type":"string","format":"","description":"OS Image reported by the node from /etc/os-release (e.g. Debian GNU/Linux 7 (wheezy)).","priority":1},{"name":"Kernel-Version","type":"string","format":"","description":"Kernel Version reported by the node from 'uname -r' (e.g. 3.16.0-0.bpo.4-amd64).","priority":1},{"name":"Container-Runtime","type":"string","format":"","description":"ContainerRuntime Version reported by the node through runtime remote API (e.g. containerd://1.4.2).","priority":1}],"rows":[{"cells":["127.0.0.1","Ready","\u003cnone\u003e","125m","v1.32.0-alpha.0.400+f0307eefb87bde-dirty","127.0.0.1","\u003cnone\u003e","Debian GNU/Linux 12 (bookworm)","6.1.0-18-amd64","cri-o://1.30.0"],"object":{"kind":"PartialObjectMetadata","apiVersion":"meta.k8s.io/v1","metadata":{"name":"127.0.0.1","uid":"7523429e-2e31-467f-98e2-1ab69b48c103","resourceVersion":"2743","creationTimestamp":"2024-09-03T06:45:10Z","labels":{"beta.kubernetes.io/arch":"amd64","beta.kubernetes.io/os":"linux","kubernetes.io/arch":"amd64","kubernetes.io/hostname":"127.0.0.1","kubernetes.io/os":"linux"},"annotations":{"node.alpha.kubernetes.io/ttl":"0","volumes.kubernetes.io/controller-managed-attach-detach":"true"},"managedFields":[{"manager":"test-client-A","operation":"Apply","apiVersion":"v1","time":"2024-09-03T06:48:38Z","fieldsType":"FieldsV1","fieldsV1":{"f:spec":{"f:taints":{"k:{\"effect\":\"NoSchedule\",\"key\":\"tainted-by-client-a\"}":{"f:value":{}}}}}},{"manager":"test-client-B","operation":"Apply","apiVersion":"v1","time":"2024-09-03T06:48:38Z","fieldsType":"FieldsV1","fieldsV1":{"f:spec":{"f:taints":{"k:{\"effect\":\"NoSchedule\",\"key\":\"tainted-by-client-b\"}":{"f:value":{}}}}}},{"manager":"kube-controller-manager","operation":"Update","apiVersion":"v1","time":"2024-09-03T06:45:10Z","fieldsType":"FieldsV1","fieldsV1":{"f:metadata":{"f:annotations":{"f:node.alpha.kubernetes.io/ttl":{}}}}},{"manager":"kubelet","operation":"Update","apiVersion":"v1","time":"2024-09-03T06:45:10Z","fieldsType":"FieldsV1","fieldsV1":{"f:metadata":{"f:annotations":{".":{},"f:volumes.kubernetes.io/controller-managed-attach-detach":{}},"f:labels":{".":{},"f:beta.kubernetes.io/arch":{},"f:beta.kubernetes.io/os":{},"f:kubernetes.io/arch":{},"f:kubernetes.io/hostname":{},"f:kubernetes.io/os":{}}}}},{"manager":"kubelet","operation":"Update","apiVersion":"v1","time":"2024-09-03T08:49:43Z","fieldsType":"FieldsV1","fieldsV1":{"f:status":{"f:conditions":{"k:{\"type\":\"DiskPressure\"}":{"f:lastHeartbeatTime":{}},"k:{\"type\":\"MemoryPressure\"}":{"f:lastHeartbeatTime":{}},"k:{\"type\":\"PIDPressure\"}":{"f:lastHeartbeatTime":{}},"k:{\"type\":\"Ready\"}":{"f:lastHeartbeatTime":{},"f:lastTransitionTime":{},"f:message":{},"f:reason":{},"f:status":{}}},"f:images":{},"f:runtimeHandlers":{}}},"subresource":"status"}]}}}]}
 >

$ _output/bin/kubectl get -v=9 nodes
I0903 10:51:56.783156  420056 helper.go:246] "Request Body" body=<
	{"apiVersion":"resource.k8s.io/v1alpha3","kind":"ResourceClaim","metadata":{"name":"example","namespace":"default"},"spec":{"devices":{"requests":[{"deviceClassName":"example","name":"req-0"}]}}}
 >
I0903 10:51:56.783293  420056 round_trippers.go:466] curl -v -XPOST  -H "Accept: application/json" -H "Content-Type: application/json" -H "User-Agent: kubectl/v1.32.0 (linux/amd64) kubernetes/7dbe980" 'https://localhost:6444/apis/resource.k8s.io/v1alpha3/namespaces/default/resourceclaims?fieldManager=kubectl-create&fieldValidation=Strict'
I0903 10:51:56.790802  420056 round_trippers.go:553] POST https://localhost:6444/apis/resource.k8s.io/v1alpha3/namespaces/default/resourceclaims?fieldManager=kubectl-create&fieldValidation=Strict 201 Created in 7 milliseconds
I0903 10:51:56.790840  420056 round_trippers.go:570] HTTP Statistics: GetConnection 0 ms ServerProcessing 7 ms Duration 7 ms
I0903 10:51:56.790855  420056 round_trippers.go:577] Response Headers:
I0903 10:51:56.790871  420056 round_trippers.go:580]     Content-Type: application/json
I0903 10:51:56.790880  420056 round_trippers.go:580]     X-Kubernetes-Pf-Flowschema-Uid: f6e769a1-a9e5-4aac-b577-5caac5bb2904
I0903 10:51:56.790894  420056 round_trippers.go:580]     X-Kubernetes-Pf-Prioritylevel-Uid: 37118ad8-6c89-4219-94d6-451d8a78dae1
I0903 10:51:56.790904  420056 round_trippers.go:580]     Content-Length: 577
I0903 10:51:56.790915  420056 round_trippers.go:580]     Date: Tue, 03 Sep 2024 08:51:56 GMT
I0903 10:51:56.790925  420056 round_trippers.go:580]     Audit-Id: 70995597-2c3b-4479-93bf-4c80d21a6f25
I0903 10:51:56.790936  420056 round_trippers.go:580]     Cache-Control: no-cache, private
I0903 10:51:56.790995  420056 helper.go:246] "Response Body" body=<
	{"kind":"ResourceClaim","apiVersion":"resource.k8s.io/v1alpha3","metadata":{"name":"example","namespace":"default","uid":"151ed596-1384-4fe4-bb64-484231f24746","resourceVersion":"2787","creationTimestamp":"2024-09-03T08:51:56Z","managedFields":[{"manager":"kubectl-create","operation":"Update","apiVersion":"resource.k8s.io/v1alpha3","time":"2024-09-03T08:51:56Z","fieldsType":"FieldsV1","fieldsV1":{"f:spec":{"f:devices":{"f:requests":{}}}}}]},"spec":{"devices":{"requests":[{"name":"req-0","deviceClassName":"example","allocationMode":"ExactCount","count":1}]}},"status":{}}
 >
resourceclaim.resource.k8s.io/example created
```

Reply:
The `helper.go` in kubectl is:
https://github.com/kubernetes/kubernetes/blob/c86a2d6925a61ac181468b74f573518db1d645d2/staging/src/k8s.io/cli-runtime/pkg/resource/helper.go#L246

Reply:
Thanks, looks good enough.